File: EduMS-Frontend/edums-frontend-DDD-System/src/app/modules/m1-school-office/interface/academic-lock-period.ts
import type { RecordStatus } from './common';

export interface AcademicLockPeriod {
  id: number;
  officeId: number;
  schoolId: number;
  periodName: string;
  startDate: string;
  endDate: string;
  isActive: boolean;
  lockGradeRosters: boolean;
  lockEnrollmentSnapshots: boolean;
  lockPeriodStatisticalReports: boolean;
  lockAttendanceLogs: boolean;
  lockBehavioralRecords: boolean;
  lockFinancialFeeAssessments: boolean;
  unlockReasonDescription: string | null;
  initiatedByEmployeeId: number | null;
}

export type CreateAcademicLockPeriodDto = Omit<AcademicLockPeriod, 'id' | 'isActive'>;

export type UpdateAcademicLockPeriodDto = Omit<AcademicLockPeriod, 'isActive' | 'schoolId'>;

File: 01_Backend/EduMS.Backend/src/EduMS.Domain/Entities/M1_SchoolAdmin/AcademicLockPeriod.cs
using EduMS.Domain.Common;

namespace EduMS.Domain.Entities;

public class AcademicLockPeriod : BaseAuditableEntity
{
    public long OfficeId { get; set; } // The office initiating the lock
    public long SchoolId { get; set; } // The school being locked
    public string PeriodName { get; set; } = string.Empty; // e.g. "2026_Semester1_FinalReviews"
    public DateTime StartDate { get; set; }
    public DateTime EndDate { get; set; }
    public bool IsActive { get; set; } = true;

    // Granular Lock Parameters (Selective locking controls)
    public bool LockGradeRosters { get; set; } = true;
    public bool LockEnrollmentSnapshots { get; set; } = true;
    public bool LockPeriodStatisticalReports { get; set; } = true;
    public bool LockAttendanceLogs { get; set; } = true;
    public bool LockBehavioralRecords { get; set; } = true;
    public bool LockFinancialFeeAssessments { get; set; } = true;
    public string? UnlockReasonDescription { get; set; }
    public long? InitiatedByEmployeeId { get; set; }

    // Cross-Module Navigation Properties
    public virtual School? School { get; set; }
    public virtual Employee? InitiatedByEmployee { get; set; }
    public virtual ICollection<StatisticalReportSnapshot> StatisticalReportSnapshots { get; set; } = new List<StatisticalReportSnapshot>();
}
